openhtf/io/output/print_console.py

```diff
@@ -0,0 +1,81 @@
+  """Return an output callback that writes JSON Test Records.
+
+  An example filename_pattern might be:
+    '/data/test_records/%(dut_id)s.%(start_time_millis)s'
+
+  To use this output mechanism:
+    test = openhtf.Test(PhaseOne, PhaseTwo)
+    test.AddOutputCallback(openhtf.OutputToJson(
+        '/data/test_records/%(dut_id)s.%(start_time_millis)s'))
+
+  Args:
+    filename_pattern: A format string specifying the filename to write to,
+      will be formatted with the Test Record as a dictionary.
+    inline_attachments: Whether attachments should be included inline in the
+      output.  Set to False if you expect to have large binary attachments.
+  """
+
+  def __init__(self, filename_pattern=None, inline_attachments=True, **kwargs):
+    super(PrintConsole, self).__init__(**kwargs)
+    self.FAIL =  '\033[91m'
+    self.PASS =  '\033[92m'
+    self.ERROR = '\033[93m'
+    self.ENDC =  '\033[0m'
+    self.BOLD =  '\033[1m'
+
+  # pylint: disable=invalid-name
+  def __call__(self, test_record):
+    as_dict = util.convert_to_dict(test_record)
+    result = as_dict['outcome']
+    color = None
+    if result == 'PASS':
+      color = self.PASS
+    elif result == 'FAIL':
+      color = self.FAIL
+    else:
+      color = self.ERROR
+
+    print '\n'+color+self.BOLD+os.path.basename(sys.argv[0])+": "+result+self.ENDC
+      
+    if result == 'FAIL' or result == 'ERROR':
+      phases=as_dict['phases']
+      for phase in phases:
+        phase_name = phase['name']
+        phase_time = str((float(phase['end_time_millis']) - \
+                               float(phase['start_time_millis']))/1000)
+        if result == 'FAIL':
+          measured = phase['measured_values']
+          measurements = phase['measurements']
+          for measurement in measurements:
+            res = measurements[measurement]
+            if res['outcome'] == 'FAIL':
+              if len(phase_name) > 0:
+                print '\nfailed phase: ' + phase_name + ' [time taken(s): ' + \
+                                                      phase_time+']'+self.ENDC
+                phase_name = ''
+
+              failed_item = res['name']
+              print ' '*4 + 'failed_item: ' + failed_item + self.ENDC
+              print ' '*8 + 'measured_value: '+ str(measured[failed_item]) + self.ENDC 
+              if 'validators' in res:
+                print ' '*8 + 'validator: ' + res['validators'][0]
+              else:
+                print ' '*8 + 'validator: None'
+        else:
+          phase_result = phase['result']
+          if phase_result['raised_exception'] == 'True':
+            print 'raised_exception phase: ' + phase_name +' [time taken(s): '\
+                              + phase_time + ']' + self.ENDC
+            exp_type = as_dict['outcome_details'][0]['code']
+            print ' '*4 + 'exception type : ' + exp_type + self.ENDC
+            print ' '*4 + 'exception text: ' + phase_result['phase_result'] + self.ENDC
```

Comment:
The typical rule of thumb is to only use print for interactive scripts (such as a REPL using `input()` and `print`).  `print` behaves slightly differently in some edge cases, and IMO is less clear about the exact output (ie, commas magically turn into space-separation, implicit newline at the end).

Also, print is less forward-compatible with Python3, and doesn't allow for easily outputting to something else (whereas it would be easy to parameterize `sys.stdout` and write to a file instead).

As a side note, I wasn't feeling super clever at 6 am, but now that I look at it, you might also benefit from a small helper class that supports indentation syntax, perhaps like:
```
class IndentingFile(object):
  def __init__(self, output_file, indent=2):
    self._output_file = output_file
    self._current_indent = 0
    self._indent = indent

  def write(self, data):
    self.output_file.write(''.join((self._current_indent * ' ', data, '\n')))

  @contextlib.contextmanager
  def indent(self):
    self._current_indent += self._indent
    yield self
    self._current_indent -= self._indent
```
Then you can:
```
def foo(test_record):
  indenter = IndentingFile(sys.stdout)
  indenter.write('first line')
  with indenter:
    indenter.write('indented one level')
    with indenter:
      indenter.write('indented two levels')
```
That may be more or less readable, just throwing the idea out there.